README.md
# antifreeze2
[![Build Status](https://travis-ci.com/DigitalBrainJS/antifreeze2.svg?branch=master)](https://travis-ci.com/DigitalBrainJS/antifreeze2)
[![](https://badgen.net/npm/license/antifreeze2)](https://unpkg.com/antifreeze2/dist/antifreeze2.umd.js)
[![Coverage Status](https://coveralls.io/repos/github/DigitalBrainJS/antifreeze2/badge.svg?branch=master)](https://coveralls.io/github/DigitalBrainJS/antifreeze2?branch=master)
[![](https://badgen.net/github/issues/DigitalBrainJS/antifreeze2)](https://github.com/DigitalBrainJS/antifreeze2/issues)
[![](https://badgen.net/github/stars/DigitalBrainJS/antifreeze2)](https://github.com/DigitalBrainJS/antifreeze2/stargazers)

:star: Antifreeze for eventloop- let it always work :star:

# Why
If you have a heavy synchronous task, you may use workers or just split the task to several async micro tasks/chunks
to keep event loop always running, its pretty easy to do that with async function.
But async function doesn't guarantee that all of its task scheduled by await will be executed really asynchronously
and will not block the IO stage of the EventLoop. This simple package consists some helpers to ensure that the event loop
is running, measuring the duration of the current event tick and allowing it go to the next tick if the maximum tick duration is exceeded.


## Installation

Install for node.js using npm/yarn:

``` bash
$ npm install antifreeze2 --save
```

``` bash
$ yarn add antifreeze2
```

````javascript
const { antifreeze, isNeeded, watchTick }= require('antifreeze2');
````

## Usage examples
#### Example 1 - Fibonacci
For example, we need to calculate Fibonacci for 1,000,000 value.
It's a task with heavy computation since it can take around 10s to complete.
If we write the function as synchronous or just use an ECMA asynchronous function, the event loop will be blocked for that period.
We won't be able to perform other tasks like accepting new connections, I/O events, timers, etc. because we only have one thread.
To avoid this, we must ensure that the event loop tick duration does not exceed the allowed range of 15-20ms in order for the application to remain responsive.

By default, the desired event loop tick is set to **10ms**. You can change it using `watchTick(maxTick: number)` function.
[See online demo](https://codesandbox.io/s/antifreeze2-basic-example-ulsvq?file=/src/index.js)
```js
import {antifreeze, isNeeded} from "antifreeze2";

// A function with heavy computations
const fibAsync = async(n) => {
  let a = 1n, b = 1n, sum, i = n - 2;
  while (i-- > 0) {
    sum = a + b;
    a = b;
    b = sum;
    if (isNeeded()) {      // If more than 10ms have passed since the last run of the eventloop cycle
      await antifreeze();  // let the event loop get polled
    }
  }
  return b;
};

// Test it - calculate Fibonacci for n= 1,000,000
(async (n) => {
  let ts = Date.now();
  let ticks = 0;

  const timer = setInterval(() => {
    const now = Date.now();
    console.log(`Timer tick [${now - ts}ms]`);
    ts = now;
    ticks++;
  }, 100);

  const result = await fibAsync(n);

  console.warn(`\nTimer ticks: ${ticks}\nFibonacci(${n}) = ${result}`)

  clearTimeout(timer);
})(500000);
```

Optionally, to get the maximum performance, you can throttle the `isNeeded()` call by using some counter:

```js
const fibAsync = async(n) => {
  let a = 1n, b = 1n, sum, i = n - 2;
  while (i-- > 0) {
    sum = a + b;
    a = b;
    b = sum;
    // check only every 1000th cycle
    if (!(i % 1000) && isNeeded()) {      // If more than 10ms have passed since the last run of the eventloop cycle
      await antifreeze();  // let the event loop get polled
    }
  }
  return b;
};
```

#### Example 2 - koa server with heavy computation

[See online demo](https://codesandbox.io/s/festive-pine-fwehi?file=/src/index.js)

The application has two endpoints:

[Time request](https://fwehi.sse.codesandbox.io/time) - light query with 20ms latency

[Fibonacci request](https://fwehi.sse.codesandbox.io/fibonacci/1000000) - heavy query that takes 10s to complete

Note that while a heavy request is being executed, the server continues to process light requests even though it is only running in one thread.

## API

<a name="module_antifreeze2"></a>

## antifreeze2

* [antifreeze2](#module_antifreeze2)
    * [.watchTick(tick)](#module_antifreeze2.watchTick)
    * [.antifreeze()](#module_antifreeze2.antifreeze) ⇒ <code>Promise.&lt;any&gt;</code> \| <code>null</code>
    * [.isNeeded([maxTick])](#module_antifreeze2.isNeeded) ⇒ <code>boolean</code>

<a name="module_antifreeze2.watchTick"></a>

### antifreeze2.watchTick(tick)
set interval for EventLoop delay checking

**Kind**: static method of [<code>antifreeze2</code>](#module_antifreeze2)  

| Param | Type | Description |
| --- | --- | --- |
| tick | <code>Number</code> | checking interval. Set to 0 to disable the watcher. By default this value is set to 15(ms) |

<a name="module_antifreeze2.antifreeze"></a>

### antifreeze2.antifreeze() ⇒ <code>Promise.&lt;any&gt;</code> \| <code>null</code>
Antifreeze promise injector

**Kind**: static method of [<code>antifreeze2</code>](#module_antifreeze2)  
<a name="module_antifreeze2.isNeeded"></a>

### antifreeze2.isNeeded([maxTick]) ⇒ <code>boolean</code>
returns true if current event loop tick is delayed

**Kind**: static method of [<code>antifreeze2</code>](#module_antifreeze2)  

| Param | Description |
| --- | --- |
| [maxTick] | max tick duration allowed |



## Contribution
Feel free to fork, open issues, enhance or create pull requests.
## License

The MIT License
Copyright (c) 2019 Dmitriy Mozgovoy <robotshara@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v0.0.2](https://github.com/DigitalBrainJS/antifreeze2/compare/v0.0.1...v0.0.2)

- Updated dependencies; [`513667e`](https://github.com/DigitalBrainJS/antifreeze2/commit/513667e632cc13fcbebb535ed8cbe14161069dd6)
- Updated README.md; [`0abb33c`](https://github.com/DigitalBrainJS/antifreeze2/commit/0abb33ce15e7cbc31f3db3b1ef5c66b522c1431d)
- Updated scripts; [`ba5b55e`](https://github.com/DigitalBrainJS/antifreeze2/commit/ba5b55e29b9d5abb366303f247afa3ddd93eab66)

#### v0.0.1

> 2 February 2022

- Initial commit [`988b0e0`](https://github.com/DigitalBrainJS/antifreeze2/commit/988b0e0b5fe520a7b7e3c7a52896fc5408f18c33)
- Updated README.md; [`519ed50`](https://github.com/DigitalBrainJS/antifreeze2/commit/519ed50b279e758697423eb5b8e47b83d34fd451)
- Added CHANGELOG.md; [`ff4e24f`](https://github.com/DigitalBrainJS/antifreeze2/commit/ff4e24ff52335ca95fd1ea1a65070d5a7db5c05c)


test/antifreeze2.spec.js
import assert from 'assert';
import {antifreeze, isNeeded} from '../lib/antifreeze2.js';

async function fibPseudoAsync(n) {
  let a = 1, b = 1, sum, i = n - 2;
  while (i-- > 0) {
    sum = a + b;
    a = b;
    b = sum;
  }
  return b;
}

async function monitorTicks(fn) {
  let max = 0;
  let ticks = 0;
  const start = process.hrtime.bigint();
  let timestamp = 0n;
  const timer = setInterval(() => {
    const now = process.hrtime.bigint();

    const duration = Number(timestamp ? now - timestamp : 0) / 1000000;
    timestamp = now;
    if (duration > max) {
      max = duration;
    }
    ticks++;
  }, 0);
  await fn();
  clearInterval(timer);
  return {
    ticks,
    max,
    duration: Number(process.hrtime.bigint() - start) / 1000000,
  }
}


describe('antifreeze', function () {
  this.timeout(0);

  it(`(HeatUp)`, async function () {
    let n = 10000;
    while (n-- > 0) {
      await fibPseudoAsync(1000);
    }
  });

  describe('eventloop', function () {
    it(`should be frozen if antifreeze is not used`, async function () {
      const result = await monitorTicks(async () => {
        let n = 10000;
        while (n-- > 0) {
          await fibPseudoAsync(10000);
        }
      });
      process.nextTick(console.log, `       Task duration ${result.duration} ms, ticks: ${result.ticks}`);
      assert.equal(result.ticks, 0);
    });
  })

  describe('isDelayed/isNeeded()', function () {
    it(`should return true if event loop tick is delayed`, async function () {
      const result = await monitorTicks(async () => {
        let n = 10000;
        while (n-- > 0) {
          await fibPseudoAsync(10000);
        }
      });
      process.nextTick(console.log, `       Task duration: ${result.duration} ms, ticks: ${result.ticks}`);
      assert.equal(isNeeded(1), true);
    })
  });

  describe('antifreeze()', function () {
    it(`should unfreeze the eventloop if it's delayed`, async function () {
      const result = await monitorTicks(async () => {
        let n = 100000;
        while (n-- > 0) {
          await fibPseudoAsync(10000);
          await antifreeze();
        }
      });
      process.nextTick(console.log, `       Task duration: ${result.duration} ms, ticks: ${result.ticks}`);
      assert.ok(result.ticks > 0, "no event loop ticks");
    });

    it(`should do nothing if the eventloop tick is not delayed enough`, async function () {
      const started = Date.now();
      const result = await monitorTicks(async () => {
        let n = 10;
        while (n-- > 0) {
          await fibPseudoAsync(10000);
          await antifreeze();
        }
      });
      process.nextTick(console.log, `       Task duration: ${result.duration} ms, ticks: ${result.ticks}`);
      assert.ok(result.ticks === 0, `event loop ticks detected [${result.ticks}]`);
      const passed = Date.now() - started;
      assert.ok(passed < 10, `the eventloop tick duration is exceeded [${passed}ms]`)
    });
  });
});


lib/antifreeze2.js
/**
 * @module antifreeze2
 */

const supportImmediate = typeof setImmediate === 'function';
const _setImmediate = supportImmediate ? setImmediate : setTimeout;

let defaultMaxTick;

const {now} = Date;

let timestamp = now();

let watcherTimer;

/**
 * set interval for EventLoop delay checking
 * @param {Number} tick - checking interval. Set to 0 to disable the watcher. By default this value is set to 15(ms)
 */

export function watchTick(tick) {
  if (tick > 0) {

    if (tick < 4) {
      throw Error('max tick can not be less than 4');
    }

    const interval = Math.max(tick, 4);

    defaultMaxTick = tick;

    if (!watcherTimer) {
      clearInterval(watcherTimer);
    }

    watcherTimer = setInterval(() => {
      timestamp = now();
    }, interval);

    if (typeof watcherTimer.unref === 'function') {
      watcherTimer.unref();
    }

  } else if (watcherTimer) {
    clearInterval(watcherTimer);
    watcherTimer = null;
  }
};

/**
 * Antifreeze promise injector
 * @returns {Promise<any>|null}
 */

export function antifreeze(value) {
  let ts = now();

  return ts - timestamp > defaultMaxTick ? new Promise(resolve => {
    timestamp = ts;
    _setImmediate(resolve, value);
  }) : null;
}

/**
 * returns true if current event loop tick is delayed
 * @param [maxTick] max tick duration allowed
 * @returns {boolean}
 */

export function isNeeded(maxTick = defaultMaxTick) {
  return now() - timestamp > maxTick;
}

watchTick(10);



example/http-app/server.js
import CPKoa from "cp-koa";
import Router from "@koa/router";
import { CPromise, CanceledError, E_REASON_CANCELED } from "c-promise2";
import assert from "assert";
import {antifreeze, isNeeded} from "../../lib/antifreeze2.js";

const fibAsync = CPromise.promisify(function* (n) {
  this.innerWeight(0);
  let a = 1n, b = 1n, sum, i = n - 2;
  while (i-- > 0) {
    sum = a + b;
    a = b;
    b = sum;
    if (isNeeded()) {
      yield antifreeze();
      this.progress(1 - (i /(n-2)));
    }
  }
  return b;
});

const app = new CPKoa();

const router = new Router();

app.use(function* (ctx, next) {
  const ts = Date.now();
  try {
    yield next();
    const passed = Date.now() - ts;
    console.warn(`Request [${ctx.req.url}] completed in [${passed}ms]`);
    ctx.set('X-Response-Time',`${passed}ms`);
  } catch (err) {
    if(CPromise.isCanceledError(err)){
      console.warn(`Request [${ctx.req.url}] canceled in [${Date.now() - ts}ms]`);
    }
    throw err;
  }
});

router.get('/', async(ctx)=>{
  ctx.body = `
  <html><body>
    <a href="/time">Current time</a><br/>
    <a href="/fibonacci/10000">fibonacci 10k</a><br/>
    <a href="/fibonacci/100000">fibonacci 100k</a><br/>
    <a href="/fibonacci/1000000">fibonacci 1M</a><br/>
    <a href="/fibonacci/2000000">fibonacci 2M</a><br/>
  </body></html>`
})

router.get("/time", async function (ctx, next) {
    const now = new Date();
    ctx.body= `Time is: ${now.toLocaleString()} (${+now})`
});

router.get("/fibonacci/:n", async function (ctx, next) {
  await ctx.run(function* () {
    let {n = 1000} = ctx.request.params;
    console.log('start calculation...');
    n *= 1;

    assert.ok(n >= 1 && n <= 10000000, `n must be in range [1..10M]`);
    const promise = fibAsync(n);

    ctx.body = `Fibonacci result for n= ${n}: ${yield promise}`;
  })
});

app
  .use(router.routes())
  .use(router.allowedMethods())
  .on("progress", (ctx, score) => {
    console.log(`Request [${ctx.req.url}] progress [${ctx.ip}]: ${(score * 100).toFixed(1)}%`);
  })
  .listen(3000);
